From c8f093d221aa4f91c372b880ff4d1b8c7dc78664 Mon Sep 17 00:00:00 2001
From: Maxime Jourdan <mjourdan@baylibre.com>
Date: Tue, 13 Nov 2018 11:13:06 +0100
Subject: [PATCH 23/23] meson: vdec: don't allow dropping timestamps with hevc

Because of how the HEVC module work, we really don't want to be dropping
timestamps that we think won't be processed.

This is a slight hack and there's most likely a better way to do it.

Fixes a few remaining issues with seeking.
---
 drivers/media/platform/meson/vdec/codec_hevc.c   | 2 +-
 drivers/media/platform/meson/vdec/vdec_helpers.c | 7 +++++--
 drivers/media/platform/meson/vdec/vdec_helpers.h | 3 +--
 3 files changed, 7 insertions(+), 5 deletions(-)

diff --git a/drivers/media/platform/meson/vdec/codec_hevc.c b/drivers/media/platform/meson/vdec/codec_hevc.c
index 7b07fef..116e9ff 100644
--- a/drivers/media/platform/meson/vdec/codec_hevc.c
+++ b/drivers/media/platform/meson/vdec/codec_hevc.c
@@ -519,7 +519,7 @@ static void codec_hevc_output_frames(struct amvdec_session *sess)
 		dev_dbg(sess->core->dev, "DONE frame poc %u; vbuf %u\n",
 			tmp->poc, tmp->vbuf->vb2_buf.index);
 		amvdec_dst_buf_done_offset(sess, tmp->vbuf, tmp->offset,
-					   V4L2_FIELD_NONE);
+					   V4L2_FIELD_NONE, false);
 		list_del(&tmp->list);
 		kfree(tmp);
 		hevc->frames_num--;
diff --git a/drivers/media/platform/meson/vdec/vdec_helpers.c b/drivers/media/platform/meson/vdec/vdec_helpers.c
index 0c0480b..e7b6651 100644
--- a/drivers/media/platform/meson/vdec/vdec_helpers.c
+++ b/drivers/media/platform/meson/vdec/vdec_helpers.c
@@ -351,7 +351,7 @@ EXPORT_SYMBOL_GPL(amvdec_dst_buf_done);
 
 void amvdec_dst_buf_done_offset(struct amvdec_session *sess,
 				struct vb2_v4l2_buffer *vbuf,
-				u32 offset, u32 field)
+				u32 offset, u32 field, bool allow_drop)
 {
 	struct device *dev = sess->core->dev_dec;
 	struct amvdec_timestamp *match = NULL;
@@ -374,6 +374,9 @@ void amvdec_dst_buf_done_offset(struct amvdec_session *sess,
 			break;
 		}
 
+		if (!allow_drop)
+			continue;
+
 		/* Delete any timestamp entry that appears before our target
 		 * (not all src packets/timestamps lead to a frame)
 		 */
@@ -415,7 +418,7 @@ void amvdec_dst_buf_done_idx(struct amvdec_session *sess,
 	}
 
 	if (offset != -1)
-		amvdec_dst_buf_done_offset(sess, vbuf, offset, field);
+		amvdec_dst_buf_done_offset(sess, vbuf, offset, field, true);
 	else
 		amvdec_dst_buf_done(sess, vbuf, field);
 }
diff --git a/drivers/media/platform/meson/vdec/vdec_helpers.h b/drivers/media/platform/meson/vdec/vdec_helpers.h
index 4d529ee..f4fdc6c 100644
--- a/drivers/media/platform/meson/vdec/vdec_helpers.h
+++ b/drivers/media/platform/meson/vdec/vdec_helpers.h
@@ -36,8 +36,7 @@ void amvdec_dst_buf_done(struct amvdec_session *sess,
 			 struct vb2_v4l2_buffer *vbuf, u32 field);
 void amvdec_dst_buf_done_offset(struct amvdec_session *sess,
 				struct vb2_v4l2_buffer *vbuf,
-				u32 offset,
-				u32 field);
+				u32 offset, u32 field, bool allow_drop);
 
 /**
  * amvdec_add_ts_reorder() - Add a timestamp to the list in chronological order
-- 
2.7.4

